spring-integration-gcp/src/main/java/org/springframework/integration/gcp/storage/GcsSession.java

```diff
@@ -0,0 +1,196 @@
+/*
+ *  Copyright 2017 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.integration.gcp.storage;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.google.cloud.WriteChannel;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.BucketInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.integration.file.remote.session.Session;
+import org.springframework.util.Assert;
+import org.springframework.util.StreamUtils;
+
+/**
+ * @author João André Martins
+ */
+public class GcsSession implements Session<BlobInfo> {
```

Comment:
I think I'm missing some context here, but having to implement the whole `Session` interface seems a bit of an overkill here. It looks like a lot of the functionality is similar to `GoogleStorageResourceObject`. For example, why is support for renaming and deleting needed to implement the channel adapters?

Reply:
This is how those GCS channel adapters are based. There is a base functionality in the `spring-integration-file` for remote files processing and it comes with the `Session` and `RemoteFileTemplate` abstractions.

I would say it is overkill to have that `GoogleStorageResourceObject` since all the storage functionality is covered by the `RemoteFileTemplate` and appropriate `Session` implementation.

Reply:
Fair enough, but in either case we should find a way to share the implantation. @joaoandremartins Can you create an issue for that?

Reply:
@meltsufin I don't think we want to create the Spring Integration dependency in the Spring Storage module. WDYT?

Reply:
I think the idea was exactly opposite: to reuse `GoogleStorageResourceObject` here instead of `GcsSession`.
But IMO that would be much harder to support in the future. As I said not one time: I would better stay away from the overkill in the `Resourse` implementation and would keep it just for simple read-only `@Value`. Everything rest can be done by special proprietary API. But you have done that already, so let's at least keep them separately to avoid more confuses. In most cases this `Session` is used from the SI Channel Adapters and there you really don't care what API is there underneath.

Reply:
Yes I'm also much more comfortable doing it the Spring Integration way here and the Resource way in the Resource land.